### ISSUE TITLE
issue on handling a great number of new defined columns

### ISSUE THREAD

**Issue by YuxiLing** [PRQL/prql#2761]
### What happened?

prql version: 0.8.1

comments: I was trying to use multiple `select`, `derive`, `take`, and `filter` operations to create a great number of new columns with various arithmetical relations. I found prql would make mistakes as the complexity or number of new columns increases. 

In the following example, generated sql uses a new column, `_expr_1`, before it is defined (line 12). However, if I change anything in this prql code, this issue will not be triggered.

I am not sure what causes this issue. 

error message from the database: column "_expr_1" does not exist

Hope this would be helpful.

### PRQL input

```elm
from tb1
sort [-c0]
select [nc1 = c1,nc2 = c2]
derive [nc3 = nc1,nc4 = nc2]
sort [-nc1]
derive [nc5 = 10/nc4]
select [
test_1 = nc1+1,
test_2 = nc1+2,
test_3 = nc1+3,
test_4 = nc1+4,
test_5 = nc1+5,
test_6 = nc1+6,
test_7 = nc1+7,
test_8 = nc1+8,
test_9 = nc1+9,
test_10 = nc5]
```

### SQL output

```SQL
WITH table_5 AS (
  SELECT
    c1 + 1 AS test_1,
    c1 + 2 AS test_2,
    c1 + 3 AS test_3,
    c1 + 4 AS test_4,
    c1 + 5 AS test_5,
    c1 + 6 AS test_6,
    c1 + 7 AS test_7,
    c1 + 8 AS test_8,
    c1 + 9 AS test_9,
    10 / _expr_1 AS test_10,
    c1 AS _expr_0,
    c2 AS _expr_1,
    c0
  FROM
    tb1
  ORDER BY
    c0 DESC
),
table_3 AS (
  SELECT
    test_1,
    test_2,
    test_3,
    test_4,
    test_5,
    test_6,
    test_7,
    test_8,
    test_9,
    test_10,
    _expr_0,
    _expr_1
  FROM
    table_5 AS table_4
),
table_2 AS (
  SELECT
    test_1,
    test_2,
    test_3,
    test_4,
    test_5,
    test_6,
    test_7,
    test_8,
    test_9,
    test_10,
    _expr_0
  FROM
    table_3 AS table_0
  ORDER BY
    _expr_0 DESC
)
SELECT
  test_1,
  test_2,
  test_3,
  test_4,
  test_5,
  test_6,
  test_7,
  test_8,
  test_9,
  test_10
FROM
  table_2 AS table_1
```

### Expected SQL output

```SQL
WITH table_5 AS (
  SELECT
    c1 + 1 AS test_1,
    c1 + 2 AS test_2,
    c1 + 3 AS test_3,
    c1 + 4 AS test_4,
    c1 + 5 AS test_5,
    c1 + 6 AS test_6,
    c1 + 7 AS test_7,
    c1 + 8 AS test_8,
    c1 + 9 AS test_9,
    c1 AS _expr_0,
    c2 AS _expr_1,
    c0
  FROM
    tb1
  ORDER BY
    c0 DESC
),
table_3 AS (
  SELECT
    test_1,
    test_2,
    test_3,
    test_4,
    test_5,
    test_6,
    test_7,
    test_8,
    test_9,
    test_10,
    _expr_0,
    _expr_1,
   10 / _expr_1 AS test_10,
  FROM
    table_5 AS table_4
  ORDER BY
    _expr_0 DESC
)
SELECT
  test_1,
  test_2,
  test_3,
  test_4,
  test_5,
  test_6,
  test_7,
  test_8,
  test_9,
  test_10
FROM
  table_3 AS table_1
```

### MVCE confirmation

- [X] Minimal example
- [X] New issue

**Comment by max-sixty** [comment]
@YuxiLing these are excellent bug reports, thank you!

**Comment by max-sixty** [comment]
I had a brief look at this, one quick observation.

Here's the query in 0.9.0 syntax:

```elm
from tb1
sort {(-c0)}
select {nc1 = c1, nc2 = c2}
derive {nc3 = nc1, nc4 = nc2}
sort {(-nc1)}
derive {nc5 = nc4 + 1}
select {
  test_1 = nc1 + 1,
  test_2 = nc1 + 2,
  test_3 = nc1 + 3,
  test_4 = nc1 + 4,
  test_5 = nc1 + 5,
  test_6 = nc1 + 6,
  test_7 = nc1 + 7,
  test_8 = nc1 + 8,
  test_9 = nc1 + 9,
  test_10 = nc5,
}
```

One quite surprising thing is that this diff in the input:

```diff
   test_6 = nc1 + 6,
   test_7 = nc1 + 7,
   test_8 = nc1 + 8,
-  test_9 = nc1 + 9,
+#   test_9 = nc1 + 9,
   test_10 = nc5,
 }
```

...causes this diff in the output:

```diff
diff --git a/2761.sql b/2761-2.sql
index 0c72aecb..ea1ce900 100644
--- a/2761.sql
+++ b/2761-2.sql
@@ -1,15 +1,5 @@
 WITH table_2 AS (
   SELECT
-    c1 + 1 AS test_1,
-    c1 + 2 AS test_2,
-    c1 + 3 AS test_3,
-    c1 + 4 AS test_4,
-    c1 + 5 AS test_5,
-    c1 + 6 AS test_6,
-    c1 + 7 AS test_7,
-    c1 + 8 AS test_8,
-    c1 + 9 AS test_9,
-    _expr_1 + 1 AS test_10,
     c1 AS _expr_0,
     c2 AS _expr_1,
     c0
@@ -18,16 +8,6 @@ WITH table_2 AS (
 ),
 table_0 AS (
   SELECT
-    test_1,
-    test_2,
-    test_3,
-    test_4,
-    test_5,
-    test_6,
-    test_7,
-    test_8,
-    test_9,
-    test_10,
     _expr_0,
     _expr_1
   FROM
@@ -35,16 +15,15 @@ table_0 AS (
 ),
 table_1 AS (
   SELECT
-    test_1,
-    test_2,
-    test_3,
-    test_4,
-    test_5,
-    test_6,
-    test_7,
-    test_8,
-    test_9,
-    test_10,
+    _expr_0 + 1 AS test_1,
+    _expr_0 + 2 AS test_2,
+    _expr_0 + 3 AS test_3,
+    _expr_0 + 4 AS test_4,
+    _expr_0 + 5 AS test_5,
+    _expr_0 + 6 AS test_6,
+    _expr_0 + 7 AS test_7,
+    _expr_0 + 8 AS test_8,
+    _expr_1 + 1 AS test_10,
     _expr_0
   FROM
     table_0
@@ -58,7 +37,6 @@ SELECT
   test_6,
   test_7,
   test_8,
-  test_9,
   test_10
 FROM
   table_1
```

Very surprising!

**Comment by aljazerzen** [comment]
Closed by #2945 or one of previous commits.

The behavior now is:

```elm
from tb1
sort {-c0}
select {nc1 = c1,nc2 = c2}
derive {nc3 = nc1,nc4 = nc2}
sort {-nc1}
derive {nc5 = 10/nc4}
select {
test_1 = nc1+1,
test_2 = nc1+2,
test_3 = nc1+3,
test_4 = nc1+4,
test_5 = nc1+5,
test_6 = nc1+6,
test_7 = nc1+7,
test_8 = nc1+8,
test_9 = nc1+9,
test_10 = nc5}
```

```sql
WITH table_0 AS (
  SELECT
    c1 + 1 AS test_1,
    c1 + 2 AS test_2,
    c1 + 3 AS test_3,
    c1 + 4 AS test_4,
    c1 + 5 AS test_5,
    c1 + 6 AS test_6,
    c1 + 7 AS test_7,
    c1 + 8 AS test_8,
    c1 + 9 AS test_9,
    (10 / c2) AS test_10,
    c1 AS _expr_0,
    c0
  FROM
    tb1
)
SELECT
  test_1,
  test_2,
  test_3,
  test_4,
  test_5,
  test_6,
  test_7,
  test_8,
  test_9,
  test_10
FROM
  table_0
ORDER BY
  _expr_0 DESC

-- Generated by PRQL compiler version:0.8.1 (https://prql-lang.org)
```

**Comment by max-sixty** [comment]
Thanks a lot!
